### DISCUSSION TITLE
Components with/without subcomponents

### DISCUSSION THREAD

**Discussion by elavy-harris** [nextgenhealthcare/connect#5141] [Q&A] [answered]
Consider the following values for PID.3:
* 123^^^^^XYZ
* 123^^^^^XYZ&

In the first case, I can get 'XYZ' as
     msg['PID']['PID.3']['PID.3.6'].toString()
In the second case, I can get 'XYZ' as
     msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString()

Neither of those expressions works in the other case, however.  The first expression returns XML for the 2nd case; the second expression returns an empty string for the 1st case.

How can I reliably get 'XYZ' without knowing which case applies?  (For the record, this does not seem to be an issue for a field with/without a component delimiter.  msg['PV1']['PV1.19']['PV1.19.1'] always works.)

Thanks.

Elliott

**Comment by pacmano1** [comment]
I try not to over think it  that much, this just tests the length:

if there is a 3.6.1 use it, other wise just use 3.6:

`$c('x', msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].length() == 1 ? msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() : msg['PID']['PID.3']['PID.3.6'].toString())`

**Reply by elavy-harris** [reply]
Thank you.  While this seems to work, the basis for the check on length() is very non-intuitive.  Despite the description in the user guide, which says that length() returns the number of child elements, it seems actually to return the number of repetitions (which is always 0 or 1 for a component or sub-component).

We've found that
```
msg['PID']['PID.3']['PID.3.6'].children().length() > 1 ? 
   msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() : 
   msg['PID']['PID.3']['PID.3.6'].toString()
```
works also, and this seems like it more clearly identifies the second case.

**Reply by odoodo** [reply]
```js
// use PID-3.6.1 or if empty use PID-3.6 instead
var content = msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() || msg['PID']['PID.3']['PID.3.6'].toString();
```

**Reply by elavy-harris** [reply]
This does not work if PID.3 looks like
     12232^^^^^&subcomp2
In that case, it returns
    <PID.3.6><PID.3.6.1/><PID.3.6.2>subcomp2</PID.3.6.2></PID.3.6>

**Reply by odoodo** [reply]
Your last example is a different requirement. This adaption should handle it:

```js
// use PID-3.6.1 or if empty use PID-3.6 instead, if there are no other sub-components
var content = msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() || (!msg['PID']['PID.3']['PID.3.6']['PID.3.6.2']  &&  msg['PID']['PID.3']['PID.3.6'].toString()) || '';
```

**Reply by elavy-harris** [reply]
No, you're focusing on the examples and not the issue.  The issue is that neither msg['PID']['PID.3']['PID.3.6'].toString() nor msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() can be used reliably _without knowing the contents of PID.3.6_.  The response by @pacmano1 provides one way to know which one will give the desired data; my response to that post provides another.

**Reply by odoodo** [reply]
mine should work, too

**Reply by elavy-harris** [reply]
Only if you know something about the contents.  There could be any number of combinations of populated/unpopulated subcomponents.

**Reply by odoodo** [reply]
If I am not heavily mistaken, this should not be the case. Try my last example with different variations. It should always return the right value for the first component/only value of a field.

**Reply by odoodo** [reply]
The Idea is 

1. Try to return the value of the required component
2. if this component does not deliver a value
3. check if there are more than 1 component
4. If not, provide the value of the field
5. Otherwise return an empty string

**Comment by tonygermano** [comment]
In your example here https://github.com/nextgenhealthcare/connect/discussions/5141#discussioncomment-2618286, you can see the xml looks like `<PID.3.6><PID.3.6.1/><PID.3.6.2>subcomp2</PID.3.6.2></PID.3.6>`. If you turn off the "parse subcomponents" option in the data type properties, it will instead look like `<PID.3.6>&amp;subcomp2</PID.3.6>`.

You could then do `content = msg['PID']['PID.3']['PID.3.6'].toString().split('&')[0] || ''`.
This should handle any of
```
123^^^^^XYZ
123^^^^^XYZ&
123^^^^^
123^^^
123^^^^^&subcomp2
```

Where you are still dealing with xml, you'll never get undefined when accessing elements by name, and toString() will give you an empty string if the element does not exist.

Once we call split on the string, we're dealing with a normal javascript array of strings. Empty subcomponents will return an empty string. Indexes past the end of the array will return `undefined`, but the or condition will return an empty string in those cases. Don't forget this array is zero based, where the subcomponents normally are numbered starting at 1.

**Reply by elavy-harris** [reply]
At least in theory, wouldn't this require me to decode any escaped characters in a subcomponent, i.e. \E\?

**Comment by elavy-harris** [comment, marked as answer]
Combining the response by @pacmano1 with my own, to create an answer:

Either of the following addresses the issue.  I prefer the latter, because I find it more intuitive.

`msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].length() == 1 ? 
    msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() : 
    msg['PID']['PID.3']['PID.3.6'].toString()`

`msg['PID']['PID.3']['PID.3.6'].children().length() > 1 ? 
   msg['PID']['PID.3']['PID.3.6']['PID.3.6.1'].toString() : 
   msg['PID']['PID.3']['PID.3.6'].toString()`

**Reply by pacmano1** [reply]
Gonna mark your own answer as the answer 'cause we could probably post another dozen ways to accomplish the same thing. lol.

You should try one with XSLT!
